source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1588,49 +1556,32 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::ClusterEntry(
   }
 }
 
-void ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::drainConnPools() {
-  for (auto& host_set : priority_set_.hostSetsPerPriority()) {
-    parent_.drainConnPools(host_set->hosts());
-  }
-}
-
-void ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::drainAllConnPools(
-    DrainConnectionsHostPredicate predicate) {
-  for (auto& host_set : priority_set_.hostSetsPerPriority()) {
-    for (const HostSharedPtr& host : host_set->hosts()) {
-      if (predicate != nullptr && !predicate(*host)) {
-        continue;
-      }
-
-      parent_.drainAllConnPoolsWorker(host);
-    }
-  }
-}
-
-void ClusterManagerImpl::ThreadLocalClusterManagerImpl::drainAllConnPoolsWorker(
-    const HostSharedPtr& host) {
-  // Drain or close any HTTP connection pool for the host. Draining an HTTP pool only leads to
-  // idle connections being closed. Non-idle connections are marked as draining and prevents new
-  // streams to go through them, causing new connections to be opened.
+void ClusterManagerImpl::ThreadLocalClusterManagerImpl::drainOrCloseConnPools(
+    const HostSharedPtr& host, absl::optional<ConnectionPool::DrainBehavior> drain_or_close) {
+  // Drain or close any HTTP connection pool for the host.
   {
     const auto container = getHttpConnPoolsContainer(host);
     if (container != nullptr) {
       container->do_not_delete_ = true;
-      container->pools_->drainConnections(
-          Envoy::ConnectionPool::DrainBehavior::DrainExistingConnections);
+      if (drain_or_close.has_value()) {
+        container->pools_->drainConnections(drain_or_close.value());
+      } else {
+        // TODO(wbpcode): in the original implementation, if the
+        // 'CLOSE_CONNECTIONS_ON_HOST_HEALTH_FAILURE' is set, the http connection pool will be
+        // drained rather then be closed. I am not sure why but just keep it.
+        container->pools_->drainConnections(
+            ConnectionPool::DrainBehavior::DrainExistingConnections);
+      }
```

Comment:
I found why we implemented this in this way.

The `CLOSE_CONNECTIONS_ON_HOST_HEALTH_FAILURE` is only supported for TCP connections. And there is no `closeConnections` interface for HTTP pools for now. I will add a comment to record it and then add a test to guard this behavior.

Reply:
OK cool thanks for the follow up!